### DISCUSSION TITLE
upgrade to phase4 - entsog profile troubles

### DISCUSSION THREAD

**Discussion by mrotaru-20** [phax/phase4#272] [Q&A]
Hi there,

I have to upgrade a legacy app to `phase4-lib` using `entsog `profile.

1. I was able to generate a mime-message (AS4MimeMessage)
But the apps use an httpclient-4.514.jar for sending the mime-message, I'm stack at HttpPost when I have to set mimeMessageEntity:

i.e. `httpPost.setEntity(HttpMimeMessageEntity.create(mime)) 
`
Certainly got a `'java.lang.ClassCastException: class com.helger.phase4.http.HttpMimeMessageEntity cannot be cast to class org.apache.http.HttpEntity'`

Does anyone have recommendations how to upgrade the client to be compatible with phase4-lib API?
The entity set works better with ``org.apache.hc.client5.http.classic.methods.HttpPost`` but this change generate further issues i.e HttpResponse.

2. When I add encryption part with _'cid:Attachments_' 
i.e. `aBuilder.getParts ().add (new WSEncryptionPart (MessageHelperMethods.PREFIX_CID + "Attachments", "Content"));`

 then I've got an exception:
` Exception was org.apache.wss4j.common.ext.WSSecurityException: http://docs.oasis-open.org/wss/oasis-wss-SwAProfile-1.1#Attachment-Content-Signature-Transform algorithm and DOM mechanism not available`

I checked the org.apache.wss4j.dom.message.WSSecSignatureBase.addReferencesToSign(WSSecSignatureBase.java:220 API and it looks for me it  is end of road, I have to discontinue using WSEncryptionPart with _cid:Attachments_ as id.

I picked the model from _https://useof.org/java-open-source/com.helger.phase4.attachment.WSS4JAttachmentCallbackHandler_   section: _the class AS4Signer method _createSignedMessage._

I'm really grateful if anyone have an advice how to use phase4-lib for entsog profile. Thanks in advance.

**Comment by phax** [comment]
Wow, okay, that sounds like an interesting topic.
Before I go into the details, please ensure the following prerequisites are fulfilled:
1. Your code is on Java 11 or later basis
2. Make sure to use the Jakarta dependencies and not the javax ones (also Angus Mail and Angus Activation)
3. Make sure to use HttpClient v5

Than start adding phase4. And then I strongly recommend to use the builder based approach instead of fiddling with the message native layouts and the http transision manually....

**Reply by mrotaru-20** [reply]
Great, thanks for quick reply :)
Right, 
I'm using Java17, I will adjust step 2 & 3 and plan to use Phase4ENTSOGSender from phase4-entsog-client-2.8.2.jar.
Hope it will fix the troubles :) 
cheers!
